WalletWasabi/ExchangeRate/ExchangeRateProvider.cs

```diff
@@ -0,0 +1,70 @@
+using System.Collections.Immutable;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using ExchangeRateProviderInfo = (string Name, bool ClearNetOnly, string ApiUrl, System.Func<string, decimal> Extractor);
+
+namespace WalletWasabi.ExchangeRate;
+
+public class ExchangeRateProvider(EndPoint? socksProxyEndPoint = null)
+{
+	public static readonly ImmutableArray<ExchangeRateProviderInfo> Providers =
+	[
+		("bitstamp", true, "https://www.bitstamp.net/api/v2/ticker/btcusd", JsonPath(".bid")),
+		("blockchain", false, "https://blockchain.info/ticker", JsonPath(".USD.buy")),
+		("coinbase", true, "https://api.coinbase.com/v2/exchange-rates?currency=BTC", JsonPath(".data.rates.USD")),
+		("coingecko", false, "https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=bitcoin", JsonPath(".[0].current_price")),
+		("coingate", true, "https://api.coingate.com/v2/rates/merchant/BTC/USD", JsonPath("$")),
+		("gemini", false, "https://api.gemini.com/v1/pubticker/btcusd", JsonPath(".bid"))
+	];
+
+	public async Task<ExchangeRate> GetExchangeRateAsync(string providerName, CancellationToken cancellationToken)
+	{
+		var providerInfo = Providers.FirstOrDefault(x => x.Name.Equals(providerName, StringComparison.InvariantCultureIgnoreCase));
+		if (providerInfo == default)
+		{
+			throw new NotSupportedException($"Exchange rate provider '{providerName}' is not supported.");
+		}
+		var proxy = GetWebProxy();
+		if (providerInfo.ClearNetOnly && proxy is not null)
+		{
+			throw new NotSupportedException($"Exchange rate provider '{providerName}' cannot be reached with Tor, only clearnet.");
+		}
+
+		var url = new Uri(providerInfo.ApiUrl);
+
+#pragma warning disable RS0030 // Do not use banned APIs
+		using var httpClientHandler = new HttpClientHandler();
+		using var httpClient = new HttpClient(httpClientHandler);
+		httpClientHandler.Proxy = proxy;
+		httpClient.BaseAddress = new Uri($"{url.Scheme}://{url.Host}");
+		httpClient.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("WasabiWallet", Helpers.Constants.ClientVersion.ToString()));
```

Comment:
Yes ofc, Kimi already saw it. Some providers require an UserAgent and we need to take random ones here.

Reply:
Yes I saw it as well. The new idea here is to replace it with random user agent to avoid identified as Wasabi. I found this website, might be useful for you. 

https://user-agents.net/random